components/ui/pathogen-map/pathogen-map-popup.tsx

```diff
@@ -0,0 +1,102 @@
+import { MapRef, Popup, useMap } from "react-map-gl";
+
+import { PathogenDataPointPropertiesBase } from "./pathogen-map";
+
+export type PopupContentGenerator<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+> = (input: TPathogenDataPointProperties) => React.ReactNode;
+
+type VisiblePopupInfo<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+> = { visible: true; properties: TPathogenDataPointProperties };
+type HiddenPopupInfo<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+> = { visible: false; properties: null };
+
+export type PopupInfo<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+> =
+  | VisiblePopupInfo<TPathogenDataPointProperties>
+  | HiddenPopupInfo<TPathogenDataPointProperties>;
+
+interface PathogenMapPopupProps<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+> {
+  mapId: string;
+  popUpInfo: PopupInfo<TPathogenDataPointProperties>;
+  generatePopupContent: PopupContentGenerator<TPathogenDataPointProperties>;
+}
+
+export function PathogenMapPopup<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+>({
+  mapId,
+  popUpInfo,
+  generatePopupContent,
+}: PathogenMapPopupProps<TPathogenDataPointProperties>) {
+  const allMaps = useMap();
+  const map = allMaps[mapId];
+
+  const getMapboxLatitudeOffset = (map: MapRef) => {
+    // Map seems to zoom in in powers of 2, so reducing offset by powers of 2 keeps the modal apprximately
+    // in the same center everytime
+    // offset needs to reduce exponentially with zoom -- higher zoom x smaller offset
+    let mapZoom = map.getZoom();
+
+    return 80 / Math.pow(2, mapZoom);
+  };
+
+  if (popUpInfo.visible === false) {
+    return null;
+  }
+
+  // Not a typo, there is some kind of underlying problem here worth investigating that
+  // makes switching the latitude and longitude a required step to produce the correct results.
+  const latitude = popUpInfo.properties.longitude ?? 0;
+  const longitude = popUpInfo.properties.latitude ?? 0;
+
+  const transformedProperties = Object.fromEntries(
+    Object.entries(popUpInfo.properties).map(([key, value]) => {
+      // Arrays in popUpInfo.properties that reach this point have been converted from
+      // ["ABC", "DEF"] to '["ABC", "DEF"]'
+      // In order to proceed further, they need to be converted back to arrays.
+      if (/^\[("[^\[\]\"]*",?)+\]$/.test(value)) {
```

Comment:
🤮 

I'm like somewhat sure this something mapbox is doing that I don't have any control over and just have to undo. Again, this is one of those questionable things I'd be happy to investigate further, I just don't want to block on it right now if that's alright.

Obviously if you disagree, just say so but again I'm pretty sure this is some mapbox stuff at work and I'm not sure how much of a say I have about this. If you're curious, the `antibodies` field is the problematic one in this case.

If you'd like to make the case that needing this suggests that my types are wrong you're probably right so I might have to ask for a pass on that one too to take another look at that when I take a look at whether the whole issue above is actually something fixable or not

Reply:
Sounds good to me, can add it in the issue from above. 

Note: antibodies has always been a problem child. Mapbox does have some issues with data points being arrays. The approach I think of here is converting the array to a string higher up and filtering on the string (changing filter logic from array.find() to string.hasSubstring() -- mehtods do not exist, only highlight logic LOL). We can work on this later though and this looks fine to me for now